spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java

```diff
@@ -0,0 +1,108 @@
+package org.baeldung.batch;
+
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.conn.ConnectTimeoutException;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.baeldung.batch.service.RetryItemProcessor;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.springframework.batch.core.ExitStatus;
+import org.springframework.batch.core.JobExecution;
+import org.springframework.batch.core.JobInstance;
+import org.springframework.batch.core.JobParameters;
+import org.springframework.batch.core.JobParametersBuilder;
+import org.springframework.batch.test.AssertFile;
+import org.springframework.batch.test.JobLauncherTestUtils;
+import org.springframework.batch.test.context.SpringBatchTest;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.core.io.FileSystemResource;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.annotation.DirtiesContext.ClassMode;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.TestExecutionListeners;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.test.context.support.DependencyInjectionTestExecutionListener;
+import org.springframework.test.context.support.DirtiesContextTestExecutionListener;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+@RunWith(SpringRunner.class)
+@SpringBatchTest
+@EnableAutoConfiguration
+@ContextConfiguration(classes = { SpringBatchRetryConfig.class })
+@TestExecutionListeners({ DependencyInjectionTestExecutionListener.class, DirtiesContextTestExecutionListener.class })
+@DirtiesContext(classMode = ClassMode.AFTER_CLASS)
+public class SpringBatchRetryIntegrationTest {
+
+    private static final String TEST_OUTPUT = "xml/retryOutput.xml";
+
+    private static final String EXPECTED_OUTPUT = "src/test/resources/output/batchRetry/retryOutput.xml";
+
+    @Autowired
+    private JobLauncherTestUtils jobLauncherTestUtils;
+
+    @Mock
+    private CloseableHttpClient httpClient;
+
+    @Mock
+    private CloseableHttpResponse httpResponse;
+
+    @InjectMocks
+    @Autowired
+    private RetryItemProcessor retryItemProcessor;
```

Comment:
This is a bit messy... we're injecting mocks AND also autowiring. Hard to follow. Does this version of spring support `@MockBean` for the `CloseableHttpClient`?

The problem you've got, though, is that the `RetryItemProcessor` creates its own http client. Really, the client should be injected via the constructor and should be defined in the `@Configuration` beans class. That would make this easier to follow.

The above does work. The `@Autowired` bean gets created by spring, injected into this class, and then line `66` does some hacking into the object to force the mocks in. It's essentially an anti-pattern for using these things together, especially as SpringTest supports mockito.

Can we tidy this up?

Reply:
I had tried using @MockBean initially but was getting NullPointerException which I couldn't resolve at that moment so I went ahead with this approach.
However, I just realized it was due to some additional annotations used in this class for which it wasn't working(nullpointer). Resolved it and have now used @MockBean.
Please review if this looks good.